homeassistant/components/homekit_controller/connection.py

```diff
@@ -75,17 +79,32 @@ def __init__(self, hass, config_entry, pairing_data):
 
         # This just tracks aid/iid pairs so we know if a HK service has been
         # mapped to a HA entity.
-        self.entities = []
+        self.entities = {}
 
         # There are multiple entities sharing a single connection - only
         # allow one entity to use pairing at once.
         self.pairing_lock = asyncio.Lock()
 
+        self.available = True
+
+    def async_set_unavailable(self):
+        """Mark state of all entities on this connection as unavailable."""
+        self.available = False
+        for entity in self.entities.values():
+            entity.async_schedule_update_ha_state()
```

Comment:
Calling entity methods outside the entity is precarious, especially for non polling entities. If a state update entity method is called before the entity has been added, it will error.

Entities are also a platform concern and it's good to keep that separation.

We prefer to use our dispatch helper to signal the correct entity method.

Reply:
Nice. Don't suppose you can suggest an integration to crib ideas on how to best use the dispatch helper from?

So right now I collect (accessory, characteristic) pairs from bridges and push them to the right entities. In this model, the connection pushes new values into the entity and then triggers a schedule_update_ha_state. The minimum change I guess would be either:

 * A dispatch helper method that took the new values and/or the availability state and the entity logic works more or less as now.
 * As above but there is a seperate dispatch helper for marking entities as unavailable.

An alternative approach which i think is a better design but a bigger change is i could store the full state on the HKDevice object (in connection.py) and when i get some new values i could trigger the entities to read the current state from connection. The entities would be responsible for updating themselves and calling schedule_update_ha_state rather than the connection being responsible (as it is on dev), but we would still avoid contention on bridges with lots of devices.

Do those make sense? Which sounds best?

Reply:
Regarding pushing state to entities or having entities pull state, there's not one correct way. Different integrations choose different paths as they see best. I think you have the best view what is best in this case.

When using the dispatch helper I would recommend having different signals, and different target methods, for different features.

Here's an example from heos:
https://github.com/home-assistant/home-assistant/blob/b9c6758dbaba368723945e3eeeb63f8811c1bcb6/homeassistant/components/heos/media_player.py#L94-L107

https://github.com/home-assistant/home-assistant/blob/b9c6758dbaba368723945e3eeeb63f8811c1bcb6/homeassistant/components/heos/__init__.py#L158-L164

Reply:
So if i understand that pattern correctly the signal would be delivered to every homekit_controller entity, not just the ones sharing the current connection. In the case of the example user with 15 bridges with 5 sensors each, i'd be dispatching to 75 entities 15 times. It should be a light weight process but i think i'd rather not do it i think.

Would it be permissible to have per-config-entry signals so i only dispatch to relevant entities? So the signal name would be something ike `DOMAIN + '_' + config_entry.id`?

Reply:
You can have as many signals as you see fit. It's good practice to create many specific signals instead of creating one signal and having to check in the target method if the signal was intended for this entity.

Reply:
Nice. I'll see what i can come up with.

Reply:
Ok i've implemented the approach where all the data is collected to a property on the connection and the dispatch function instructs the entity to update its own state. Plumbed through the config entry upload cleanup as well.